processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java

```diff
@@ -122,14 +122,13 @@ public Assignment getTargetAssignment(Method mappingMethod, String mappedElement
         ResolvingAttempt attempt = new ResolvingAttempt(
             sourceModel,
             mappingMethod,
-            mappedElement,
             dateFormat,
             numberFormat,
             sourceRHS,
             criteria
         );
 
-        return attempt.getTargetAssignment( sourceRHS.getSourceType(), targetType );
+        return attempt.getTargetAssignment( sourceRHS.getSourceTypeForMatching(), targetType );
```

Comment:
No the idea is great.

>In the old solution, the SourceType in SourceRHS was simply replaced by the type of the collection element when an ADDER was applicable.

This is what was missing in my understanding. I couldn't figure out how it was before.